netket/models/equivariant.py

```diff
@@ -24,54 +24,234 @@
+    Numpy/Jax arrays must be wrapped into an :class:`netket.utils.HashableArray`.
     """
+    irreps: Tuple[HashableArray]
+    """List of irreducible represenation matrices"""
+    layers: int
+    """Number of layers (not including sum layer over output)."""
+    features: Tuple
+    """Number of features in each layer starting from the input. If a single number is given,
+    all layers will have the same number of features."""
+    characters: HashableArray
+    """Array specifying the characters of the desired symmetry representation"""
+    dtype: Any = np.float64
+    """The dtype of the weights."""
+    activation: Any = jax.nn.selu
+    """The nonlinear activation function between hidden layers."""
+    output_activation: Any = None
+    """The nonlinear activation before the output."""
+    imag_part: bool = False
+    """If true return only the imaginary part of the output"""
+    use_bias: bool = True
+    """if True uses a bias in all layers."""
+    precision: Any = None
+    """numerical precision of the computation see `jax.lax.Precision`for details."""
+    kernel_init: NNInitFunc = unit_normal_scaling
+    """Initializer for the Dense layer matrix."""
+    bias_init: NNInitFunc = zeros
+    """Initializer for the hidden bias."""
 
-    symmetries: Union[HashableArray, PermutationGroup]
-    """A group of symmetry operations (or array of permutation indices) over which the layer should be invariant.
+    def setup(self):
 
+        self.n_symm = np.asarray(self.symmetries).shape[0]
+
+        self.dense_symm = DenseSymmMatrix(
+            symmetries=self.symmetries,
+            features=self.features[0],
+            dtype=self.dtype,
+            use_bias=self.use_bias,
+            kernel_init=self.kernel_init,
+            bias_init=self.bias_init,
+            precision=self.precision,
+        )
+
+        self.equivariant_layers = [
+            DenseEquivariantIrrep(
+                irreps=self.irreps,
+                in_features=self.features[layer],
+                out_features=self.features[layer + 1],
+                use_bias=self.use_bias,
+                dtype=self.dtype,
+                precision=self.precision,
+                kernel_init=self.kernel_init,
+                bias_init=self.bias_init,
+            )
+            for layer in range(self.layers - 1)
+        ]
+
+    @nn.compact
+    def __call__(self, x):
+        x = self.dense_symm(x)
+
+        for layer in range(self.layers - 1):
+
+            x = self.activation(x)
+            x = self.equivariant_layers[layer](x)
+
+        if self.output_activation is not None:
+            x = self.output_activation(x)
+
+        x = x.reshape(-1, self.features[-1] * self.n_symm)
+        x_max = jnp.max(x, axis=-1, keepdims=True)
+        x = jnp.exp(x - x_max)
+        x = x.reshape(-1, self.features[-1], self.n_symm)
+        x = jnp.sum(x, 1)
```

Comment:
is it possible to rewrite this stuff and what follows as a `logsumexp` to increase precision?
BTW, I recently PR'd `jax.scipy.special.logsumexp` for complex numbers to jax (you just need to update it) so you could use that if it's possible

Reply:
@chrisrothUT ☝️?